# Hey Good Morning Guys,

## Here is the assignment part 1 with basic level while go day wise level will increase.

### Where in first part we have worked on small dataset to analyze how effective is encoding while building any model & this all dataset our open sourced & from wednesday will increase the level with custom dataset.

# Before you apply any Encoding Technique,you shoul think like::
### 1.) Why Label Encoding?
### 2.) When & where to use?
### 3.) Limitation of Label Encoding?
### 4.) Advantage of Label Encoding?

## You should ask from yourself with every topic tha you are applying to make it clear with depth & practice at the same time.

In [21]:
# import all the necessary packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder

In [28]:
# Define the headers since the data does not have any

headers = ["symboling", "normalized_losses", "make", "fuel_type", "aspiration",
           "num_doors", "body_style", "drive_wheels", "engine_location",
           "wheel_base", "length", "width", "height", "curb_weight",
           "engine_type", "num_cylinders", "engine_size", "fuel_system",
           "bore", "stroke", "compression_ratio", "horsepower", "peak_rpm",
           "city_mpg", "highway_mpg", "price"]

# Read in the CSV file and convert "?" to NaN

df = pd.read_csv("imports-85.data.csv", header=None, names=headers, na_values="?")
df.head()

symboling  normalized_losses         make fuel_type aspiration num_doors  \
0          3                NaN  alfa-romero       gas        std       two   
1          3                NaN  alfa-romero       gas        std       two   
2          1                NaN  alfa-romero       gas        std       two   
3          2              164.0         audi       gas        std      four   
4          2              164.0         audi       gas        std      four   

    body_style drive_wheels engine_location  wheel_base  length  width  \
0  convertible          rwd           front        88.6   168.8   64.1   
1  convertible          rwd           front        88.6   168.8   64.1   
2    hatchback          rwd           front        94.5   171.2   65.5   
3        sedan          fwd           front        99.8   176.6   66.2   
4        sedan          4wd           front        99.4   176.6   66.4   

   height  curb_weight engine_type num_cylinders  engine_size fuel_system  \
0    48.8         2548        dohc          four          130        mpfi   
1    48.8         2548        dohc          four          130        mpfi   
2    52.4         2823        ohcv           six          152        mpfi   
3    54.3         2337         ohc          four          109        mpfi   
4    54.3         2824         ohc          five          136        mpfi   

   bore  stroke  compression_ratio  horsepower  peak_rpm  city_mpg  \
0  3.47    2.68                9.0       111.0    5000.0        21   
1  3.47    2.68                9.0       111.0    5000.0        21   
2  2.68    3.47                9.0       154.0    5000.0        19   
3  3.19    3.40               10.0       102.0    5500.0        24   
4  3.19    3.40                8.0       115.0    5500.0        18   

   highway_mpg    price  
0           27  13495.0  
1           27  16500.0  
2           26  16500.0  
3           30  13950.0  
4           22  17450.0

In [29]:
df.shape

(205, 26)

In [30]:
df.info() # check info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized_losses  164 non-null    float64
 2   make               205 non-null    object 
 3   fuel_type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num_doors          203 non-null    object 
 6   body_style         205 non-null    object 
 7   drive_wheels       205 non-null    object 
 8   engine_location    205 non-null    object 
 9   wheel_base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb_weight        205 non-null    int64  
 14  engine_type        205 non-null    object 
 15  num_cylinders      205 non-null    object 
 16  engine_size        205 non

In [31]:
# check the null value
df.isnull().sum()

symboling             0
normalized_losses    41
make                  0
fuel_type             0
aspiration            0
num_doors             2
body_style            0
drive_wheels          0
engine_location       0
wheel_base            0
length                0
width                 0
height                0
curb_weight           0
engine_type           0
num_cylinders         0
engine_size           0
fuel_system           0
bore                  4
stroke                4
compression_ratio     0
horsepower            2
peak_rpm              2
city_mpg              0
highway_mpg           0
price                 4
dtype: int64

In [32]:
new_data = df.dropna(axis=0, how='any') # drop the null  values 
new_data

symboling  normalized_losses   make fuel_type aspiration num_doors  \
3            2              164.0   audi       gas        std      four   
4            2              164.0   audi       gas        std      four   
6            1              158.0   audi       gas        std      four   
8            1              158.0   audi       gas      turbo      four   
10           2              192.0    bmw       gas        std       two   
..         ...                ...    ...       ...        ...       ...   
200         -1               95.0  volvo       gas        std      four   
201         -1               95.0  volvo       gas      turbo      four   
202         -1               95.0  volvo       gas        std      four   
203         -1               95.0  volvo    diesel      turbo      four   
204         -1               95.0  volvo       gas      turbo      four   

    body_style drive_wheels engine_location  wheel_base  length  width  \
3        sedan          fwd           front        99.8   176.6   66.2   
4        sedan          4wd           front        99.4   176.6   66.4   
6        sedan          fwd           front       105.8   192.7   71.4   
8        sedan          fwd           front       105.8   192.7   71.4   
10       sedan          rwd           front       101.2   176.8   64.8   
..         ...          ...             ...         ...     ...    ...   
200      sedan          rwd           front       109.1   188.8   68.9   
201      sedan          rwd           front       109.1   188.8   68.8   
202      sedan          rwd           front       109.1   188.8   68.9   
203      sedan          rwd           front       109.1   188.8   68.9   
204      sedan          rwd           front       109.1   188.8   68.9   

     height  curb_weight engine_type num_cylinders  engine_size fuel_system  \
3      54.3         2337         ohc          four          109        mpfi   
4      54.3         2824         ohc          five          136        mpfi   
6      55.7         2844         ohc          five          136        mpfi   
8      55.9         3086         ohc          five          131        mpfi   
10     54.3         2395         ohc          four          108        mpfi   
..      ...          ...         ...           ...          ...         ...   
200    55.5         2952         ohc          four          141        mpfi   
201    55.5         3049         ohc          four          141        mpfi   
202    55.5         3012        ohcv           six          173        mpfi   
203    55.5         3217         ohc           six          145         idi   
204    55.5         3062         ohc          four          141        mpfi   

     bore  stroke  compression_ratio  horsepower  peak_rpm  city_mpg  \
3    3.19    3.40               10.0       102.0    5500.0        24   
4    3.19    3.40                8.0       115.0    5500.0        18   
6    3.19    3.40                8.5       110.0    5500.0        19   
8    3.13    3.40                8.3       140.0    5500.0        17   
10   3.50    2.80                8.8       101.0    5800.0        23   
..    ...     ...                ...         ...       ...       ...   
200  3.78    3.15                9.5       114.0    5400.0        23   
201  3.78    3.15                8.7       160.0    5300.0        19   
202  3.58    2.87                8.8       134.0    5500.0        18   
203  3.01    3.40               23.0       106.0    4800.0        26   
204  3.78    3.15                9.5       114.0    5400.0        19   

     highway_mpg    price  
3             30  13950.0  
4             22  17450.0  
6             25  17710.0  
8             20  23875.0  
10            29  16430.0  
..           ...      ...  
200           28  16845.0  
201           25  19045.0  
202           23  21485.0  
203           27  22470.0  
204           25  22625.0  

[159 rows x 26 columns]

In [33]:
new_data.shape

(159, 26)

- **columns reduced from 205 to 159 after droping NaN's**

In [34]:
y = new_data['price']   # dependent variables
X = new_data.drop('price', axis=1)  # independent variables

In [35]:
X.shape, y.shape  # check the shape of X and y

((159, 25), (159,))

In [36]:
# split the data into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42) 

In [38]:
# check shape of train and test data
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(127, 25) (127,)
(32, 25) (32,)


### Why do we use encoding technique ?

- We use **encoding techniques** as machine learning models **won't take the raw string as input.** If we try to run the Linear regression model here **without any encoding we will get errors.** Let's try to run without encoding and see what result will we get.

In [39]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)

ValueError: could not convert string to float: 'nissan'

- We have got **error** saying **could not convert string to float.** So before giving the varibales to the model we have to encode them

In [40]:
# select all the categorical columns
obj_df = new_data.select_dtypes(include=['object']).copy()
obj_df.head()

make fuel_type aspiration num_doors body_style drive_wheels  \
3   audi       gas        std      four      sedan          fwd   
4   audi       gas        std      four      sedan          4wd   
6   audi       gas        std      four      sedan          fwd   
8   audi       gas      turbo      four      sedan          fwd   
10   bmw       gas        std       two      sedan          rwd   

   engine_location engine_type num_cylinders fuel_system  
3            front         ohc          four        mpfi  
4            front         ohc          five        mpfi  
6            front         ohc          five        mpfi  
8            front         ohc          five        mpfi  
10           front         ohc          four        mpfi

In [41]:
obj_df.shape

(159, 10)

In [42]:
# We will use One hot encoding to convert all the categorical features
# Here the biggest mistake we are making is just applying one hot encoding blindly without figuring out ordinal & nominal feature.

dum = pd.get_dummies(obj_df,drop_first=True)
dum.head()

make_bmw  make_chevrolet  make_dodge  make_honda  make_jaguar  make_mazda  \
3          0               0           0           0            0           0   
4          0               0           0           0            0           0   
6          0               0           0           0            0           0   
8          0               0           0           0            0           0   
10         1               0           0           0            0           0   

    make_mercedes-benz  make_mitsubishi  make_nissan  make_peugot  \
3                    0                0            0            0   
4                    0                0            0            0   
6                    0                0            0            0   
8                    0                0            0            0   
10                   0                0            0            0   

    make_plymouth  make_porsche  make_saab  make_subaru  make_toyota  \
3               0             0          0            0            0   
4               0             0          0            0            0   
6               0             0          0            0            0   
8               0             0          0            0            0   
10              0             0          0            0            0   

    make_volkswagen  make_volvo  fuel_type_gas  aspiration_turbo  \
3                 0           0              1                 0   
4                 0           0              1                 0   
6                 0           0              1                 0   
8                 0           0              1                 1   
10                0           0              1                 0   

    num_doors_two  body_style_hardtop  body_style_hatchback  body_style_sedan  \
3               0                   0                     0                 1   
4               0                   0                     0                 1   
6               0                   0                     0                 1   
8               0                   0                     0                 1   
10              1                   0                     0                 1   

    body_style_wagon  drive_wheels_fwd  drive_wheels_rwd  engine_type_l  \
3                  0                 1                 0              0   
4                  0                 0                 0              0   
6                  0                 1                 0              0   
8                  0                 1                 0              0   
10                 0                 0                 1              0   

    engine_type_ohc  engine_type_ohcf  engine_type_ohcv  num_cylinders_five  \
3                 1                 0                 0                   0   
4                 1                 0                 0                   1   
6                 1                 0                 0                   1   
8                 1                 0                 0                   1   
10                1                 0                 0                   0   

    num_cylinders_four  num_cylinders_six  num_cylinders_three  \
3                    1                  0                    0   
4                    0                  0                    0   
6                    0                  0                    0   
8                    0                  0                    0   
10                   1                  0                    0   

    fuel_system_2bbl  fuel_system_idi  fuel_system_mfi  fuel_system_mpfi  \
3                  0                0                0                 1   
4                  0                0                0                 1   
6                  0                0                0                 1   
8                  0                0                0                 1   
10                 0                0                0                

- **No of categorical columns increased from 10 to 39 with OneHotEncoding**

In [43]:
# drop the categorical column from new data
new_df = new_data.drop(obj_df.columns,axis=1)
new_df.head()

symboling  normalized_losses  wheel_base  length  width  height  \
3           2              164.0        99.8   176.6   66.2    54.3   
4           2              164.0        99.4   176.6   66.4    54.3   
6           1              158.0       105.8   192.7   71.4    55.7   
8           1              158.0       105.8   192.7   71.4    55.9   
10          2              192.0       101.2   176.8   64.8    54.3   

    curb_weight  engine_size  bore  stroke  compression_ratio  horsepower  \
3          2337          109  3.19     3.4               10.0       102.0   
4          2824          136  3.19     3.4                8.0       115.0   
6          2844          136  3.19     3.4                8.5       110.0   
8          3086          131  3.13     3.4                8.3       140.0   
10         2395          108  3.50     2.8                8.8       101.0   

    peak_rpm  city_mpg  highway_mpg    price  
3     5500.0        24           30  13950.0  
4     5500.0        18           22  17450.0  
6     5500.0        19           25  17710.0  
8     5500.0        17           20  23875.0  
10    5800.0        23           29  16430.0

In [44]:
new_df.shape

(159, 16)

In [45]:
# concat dum with new_df
new_dum = pd.concat([new_df,dum],axis=1)
new_dum.head()

symboling  normalized_losses  wheel_base  length  width  height  \
3           2              164.0        99.8   176.6   66.2    54.3   
4           2              164.0        99.4   176.6   66.4    54.3   
6           1              158.0       105.8   192.7   71.4    55.7   
8           1              158.0       105.8   192.7   71.4    55.9   
10          2              192.0       101.2   176.8   64.8    54.3   

    curb_weight  engine_size  bore  stroke  compression_ratio  horsepower  \
3          2337          109  3.19     3.4               10.0       102.0   
4          2824          136  3.19     3.4                8.0       115.0   
6          2844          136  3.19     3.4                8.5       110.0   
8          3086          131  3.13     3.4                8.3       140.0   
10         2395          108  3.50     2.8                8.8       101.0   

    peak_rpm  city_mpg  highway_mpg    price  make_bmw  make_chevrolet  \
3     5500.0        24           30  13950.0         0               0   
4     5500.0        18           22  17450.0         0               0   
6     5500.0        19           25  17710.0         0               0   
8     5500.0        17           20  23875.0         0               0   
10    5800.0        23           29  16430.0         1               0   

    make_dodge  make_honda  make_jaguar  make_mazda  make_mercedes-benz  \
3            0           0            0           0                   0   
4            0           0            0           0                   0   
6            0           0            0           0                   0   
8            0           0            0           0                   0   
10           0           0            0           0                   0   

    make_mitsubishi  make_nissan  make_peugot  make_plymouth  make_porsche  \
3                 0            0            0              0             0   
4                 0            0            0              0             0   
6                 0            0            0              0             0   
8                 0            0            0              0             0   
10                0            0            0              0             0   

    make_saab  make_subaru  make_toyota  make_volkswagen  make_volvo  \
3           0            0            0                0           0   
4           0            0            0                0           0   
6           0            0            0                0           0   
8           0            0            0                0           0   
10          0            0            0                0           0   

    fuel_type_gas  aspiration_turbo  num_doors_two  body_style_hardtop  \
3               1                 0              0                   0   
4               1                 0              0                   0   
6               1                 0              0                   0   
8               1                 1              0                   0   
10              1                 0              1                   0   

    body_style_hatchback  body_style_sedan  body_style_wagon  \
3                      0                 1                 0   
4                      0                 1                 0   
6                      0                 1                 0   
8                      0                 1                 0   
10                     0                 1                 0   

    drive_wheels_fwd  drive_wheels_rwd  engine_type_l  engine_type_ohc  \
3                  1                 0              0                1   
4                  0                 0              0                1   
6                  1                 0              0                1   
8                  1                 0              0                1   
10                 0                 1              0                1   

    engine_type_ohcf  engine_type_ohcv  n

In [46]:
y = new_dum['price']   # dependent variables
X = new_dum.drop('price',axis=1)  # independent variables

In [47]:
# split into training and testing
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42) 

In [48]:
# fit the regression model 
model.fit(X_train, y_train)
predict = model.predict(X_test)
model.score(X_test, y_test)

0.8080831831623196

In [49]:
model.score(X_train, y_train)

0.9807110522678218

### We have got 80% accuracy Can we prove our model performance by encoding ?

- If you notice in dataset we have applied **one hot encoding on all the features**, but we use one hot encoding only for **nominal data**.


- In our dataset we have feature column **num_cylinders (Number of cylinders)**. It is a **ordinal variable**, let's **apply label encoding** on it and see if we can improve our accuracy.

# What we learn in this notebook?

## Is that we have to figure out ordinal & nominal feature to apply accurate Encoding Technique.

In [50]:
# apply label encoding on num_cylinders

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
label = le.fit_transform(obj_df['num_cylinders'])

# drop older dum_cylinders from dum

dum_lab = dum.drop(dum[['num_cylinders_five', 'num_cylinders_four','num_doors_two',
                        'num_cylinders_six', 'num_cylinders_three']],axis=1)

# add the newly LabelEncoded num_cylinders column

dum_lab['num_cylinders'] = label

In [51]:
# concat dum_lab with new_df
new_dum_lab = pd.concat([new_df,dum_lab],axis=1)
new_dum.head()

symboling  normalized_losses  wheel_base  length  width  height  \
3           2              164.0        99.8   176.6   66.2    54.3   
4           2              164.0        99.4   176.6   66.4    54.3   
6           1              158.0       105.8   192.7   71.4    55.7   
8           1              158.0       105.8   192.7   71.4    55.9   
10          2              192.0       101.2   176.8   64.8    54.3   

    curb_weight  engine_size  bore  stroke  compression_ratio  horsepower  \
3          2337          109  3.19     3.4               10.0       102.0   
4          2824          136  3.19     3.4                8.0       115.0   
6          2844          136  3.19     3.4                8.5       110.0   
8          3086          131  3.13     3.4                8.3       140.0   
10         2395          108  3.50     2.8                8.8       101.0   

    peak_rpm  city_mpg  highway_mpg    price  make_bmw  make_chevrolet  \
3     5500.0        24           30  13950.0         0               0   
4     5500.0        18           22  17450.0         0               0   
6     5500.0        19           25  17710.0         0               0   
8     5500.0        17           20  23875.0         0               0   
10    5800.0        23           29  16430.0         1               0   

    make_dodge  make_honda  make_jaguar  make_mazda  make_mercedes-benz  \
3            0           0            0           0                   0   
4            0           0            0           0                   0   
6            0           0            0           0                   0   
8            0           0            0           0                   0   
10           0           0            0           0                   0   

    make_mitsubishi  make_nissan  make_peugot  make_plymouth  make_porsche  \
3                 0            0            0              0             0   
4                 0            0            0              0             0   
6                 0            0            0              0             0   
8                 0            0            0              0             0   
10                0            0            0              0             0   

    make_saab  make_subaru  make_toyota  make_volkswagen  make_volvo  \
3           0            0            0                0           0   
4           0            0            0                0           0   
6           0            0            0                0           0   
8           0            0            0                0           0   
10          0            0            0                0           0   

    fuel_type_gas  aspiration_turbo  num_doors_two  body_style_hardtop  \
3               1                 0              0                   0   
4               1                 0              0                   0   
6               1                 0              0                   0   
8               1                 1              0                   0   
10              1                 0              1                   0   

    body_style_hatchback  body_style_sedan  body_style_wagon  \
3                      0                 1                 0   
4                      0                 1                 0   
6                      0                 1                 0   
8                      0                 1                 0   
10                     0                 1                 0   

    drive_wheels_fwd  drive_wheels_rwd  engine_type_l  engine_type_ohc  \
3                  1                 0              0                1   
4                  0                 0              0                1   
6                  1                 0              0                1   
8                  1                 0              0                1   
10                 0                 1              0                1   

    engine_type_ohcf  engine_type_ohcv  n

In [52]:
y = new_dum_lab['price']   # dependent variables
X = new_dum_lab.drop('price', axis=1)  # independent variables

In [53]:
# split inot training and testing
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42) 

In [54]:
# fit the regression model 
model.fit(X_train, y_train)
predict = model.predict(X_test)
model.score(X_test, y_test)

0.8183410734465217

In [55]:
model.score(X_train, y_train)

0.9790660485669549

- Our Accuracy has increased **from 80% to 81% by using label encoding**. We should be always careful with the encoding technique we choose as by using right encoding method we can increase the accuracy of our model